include/GafferOSL/ShadingEngine.h

```diff
@@ -79,8 +79,22 @@ class ShadingEngine : public IECore::RefCounted
 
 		IECore::CompoundDataPtr shade( const IECore::CompoundData *points, const Transforms &transforms = Transforms() ) const;
 
+		struct AttributeRead
+		{
+			AttributeRead() {}
+			AttributeRead(const std::string& scope, const std::string& name) : scope(scope), name(name) {}
+			IECore::InternedString scope;
+			IECore::InternedString name;
+		};
+
+		typedef std::vector<AttributeRead> AttributeReadContainer;
```

Comment:
I wonder if a vector is the right container here? It seems like one common requirement will be to quickly ask "is this attribute needed?", in which case a `boost::flat_set()` might be the best choice - it has very fast lookup and slower insertion (ideal for one-time initialisation followed by many reads).